### PR TITLE
z level fall/movement refactor and qol (I'm looking at you, Tramstation!!)

### LINKED ISSUE #47376: Descending to a lower z-level and dragging something breaks grab.
Imagine, not testing the both ways. Honk

### LINKED ISSUE #49968: Can't fireman carry someone upstairs
Issue reported from  Round ID: [132255](https://scrubby.melonmesa.com/round/132255) (/tg/Station Manuel [ENGLISH] [US-WEST] [ROLEPLAY] [99% LAG FREE])

Reporting client version: 513.1513

You leave them at the top of the stairs, but the offhand symbol remains in your hand slot until you drop it.

### LINKED ISSUE #53713: Ice-Botany is not wheelchair accessible
## Round ID:
Any round after the updated botany on Icebox

## Reproduction:
Not entirely sure if this would be a "bug", but more so an oversight. The new Ice-botany doesn't allow those in wheelchairs to access the bottom level, without painfully having to hurl themselves down the stairs, or having the engineers wire up a quantumpad. 
Honestly, this may not even be a super big issue considering how little others play with the paraplegic trait, at least, seriously, that is.

### LINKED ISSUE #57711: You can't drag people "up" or "down" in space
## Reproduction:

If you've got a body on a lower z level of space, and you're using a jetpack to go "up" or "down", you can't drag the body upwards in any way.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Purging some redundant procs related to z level movement while standardizing others.  Buckled mobs and pulled movables should now be carriable down/up stairs without too much copypasta. 
The code is pretty scrappy all-nighter atm (though tested, otherwise I'd not be making the PR yet), and I'm certain changes will be requested, so I'll mark this as a draft.

## Why It's Good For The Game
Will fix #49968, fix #47376, fix #57711, fix #53713.

## Changelog
:cl:
refactor: Refactored and standardized z level fall/movement code a little.
qol: You can now drag pulled movables upstairs or downstairs (falling included; be careful of conga lining around pitfalls).
fix: Same deal with fireman carried mobs, or any movable with mobs buckled on it for that matter. Just think twice before falling down one floor while skateboarding, non-free runners.
/:cl:

I ought to improve the PR description and some of the doc-comments later.

### CODE REVIEW

**Comment by Time-Green** [comment]
Does this also work for ladders?

**Comment by Ghommie** [comment]
> 
> 
> Does this also work for ladders?

Yes, they all use the `zMove()` proc (name should be changed to be all lowercase later).

**Comment by moocowswag** [comment]
Will you be able to carry someone and fly up a z level (either wings or no gravity) instead of just dropping them?

**Comment by Ghommie** [comment]
> 
> 
> Will you be able to carry someone and fly up a z level (either wings or no gravity) instead of just dropping them?

Yes, but only if they are buckled to you.

**Comment by Jolly-66** [comment]
I wonder if this'll fix this issue too, #53713

**Comment by Qustinnus** [comment]
no, thats a mapping issue.

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -164,23 +151,62 @@
...
+			return FALSE
+		if(!can_z_move(dir, get_turf(src), target, forced ? NONE : ZTRAVEL_CAN_FLY_CHECKS))
+			if(feedback)
+				to_chat(src, "<span class='warning'>You couldn't move there!</span>")
+			return FALSE
+	var/list/fallen_heroes = list(src)
+	if(affect_pulling && pulling)
+		fallen_heroes += pulling
+		if(pulling.buckled_mobs)
+			fallen_heroes += pulling.buckled_mobs
```
Comment (MrMelbert) [inline code/game/atoms_movable.dm:171]: `Pulling a mob who's buckled to a something` vs `pulling something with a mob buckled to it` behaves slightly differently, do both situations work with this?

**Comment by Jolly-66** [comment]
> no, thats a mapping issue.

Sad.

**Review by MrMelbert — COMMENTED** [review]
Damn, I was working on my own z-movement refactor at the same time, this looks much more solid though.

File: code/game/atoms_movable.dm (hunk 1 above)
Comment (Ghommie) [reply code/game/atoms_movable.dm:171]: Umh, nice consideration. I'll give it a test and ascertain they both work equally fine.

**Comment by Zytolg** [comment]
Tram makes me happy because it makes previously existent (but rare due to not being encountered) problems, oversights, or bugs visible. Great work, and I'm looking forwards to seeing whatever other multi-z issues are dug up by tram, and fixed by PR's like this one. 👍

**Comment by cacogen** [comment]
I wish you could still bodyslam people from upper floors

**Comment by Ghommie** [comment]
> 
> 
> I wish you could still bodyslam people from upper floors

Me too, but that's out of scope.

File: code/datums/elements/entered_open_space_fall.dm [hunk 2]
```diff
@@ -0,0 +1,42 @@
+/*
+ * An element specific to turfs that can drop movables on the z level beneath when moved onto.
+ * Basically, we want the fall to happen in [/atom/movable/proc/Moved()], not in [/turf/Exited()].
+ * Should make pulling, riding vehicles etc. through z levels a bit easier without resorting to many hacks and snowflakes.
+ */
+/datum/element/entered_open_space_fall
```
Comment (Qustinnus) [inline code/datums/elements/entered_open_space_fall.dm:6]: this name is a bit weird, but I don't necesarily have a good name either. this one is a bit too over-complex while not actually explaining what the element does however.

File: code/__DEFINES/dcs/signals.dm [hunk 3]
```diff
@@ -342,11 +342,12 @@
 
 // /atom/movable signals
 
-///from base of atom/movable/Moved(): (/atom)
+///from base of atom/movable/Move(): (/atom)
 #define COMSIG_MOVABLE_PRE_MOVE "movable_pre_move"
 	#define COMPONENT_MOVABLE_BLOCK_PRE_MOVE (1<<0)
 ///from base of atom/movable/Moved(): (/atom, dir)
 #define COMSIG_MOVABLE_MOVED "movable_moved"
+	#define MOVABLE_MOVED_ZFELL_DOWN (2<<0) // Starts from here because Move() also uses bools for return value (Moved() is called only if the value is true/non-zero, so it won't screw things up)
```
Comment (Qustinnus) [inline code/__DEFINES/dcs/signals.dm:350]: That's bad. It should either be a bitflag or a bool, not both.  If you want to return a true-ey value on moved then check the actual flag thas returne,d if the result is that they fell down, then return true.

File: code/game/atoms_movable.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/game/atoms_movable.dm:171]: do they?

Comment (Ghommie) [reply code/game/atoms_movable.dm:171]: Yes, but on the local working tree, not the latest pushed commit here. I shall update this PR this weekend when I have more free time.

**Pull request closed without merging**